wagtail/core/collectors.py

```diff
@@ -0,0 +1,379 @@
+        for obj in self.model._default_manager.using(self.using) \
+                .filter(filters):
+            for field in self.fields:
+                for found_obj in find_objects_in_rich_text(
+                        getattr(obj, field.attname)):
+                    if not searched_objects or \
+                            get_obj_base_key(found_obj) in searched_data:
+                        yield obj, found_obj
+
+
+class StreamFieldCollector:
+    def __init__(self, field):
+        self.field = field
+
+    def block_tree_paths(self, block, ancestors=()):
+        if isinstance(block, (StreamBlock, StructBlock)):
+            for child_block in block.child_blocks.values():
+                yield from self.block_tree_paths(child_block,
+                                                 ancestors + (block,))
+        elif isinstance(block, ListBlock):
+            yield from self.block_tree_paths(block.child_block,
+                                             ancestors + (block,))
+        else:
+            yield ancestors + (block,)
+
+    def find_block_type(self, block_types):
+        for block_path in self.block_tree_paths(self.field.stream_block):
+            if isinstance(block_path[-1], block_types):
+                yield block_path
+
+    def find_values(self, stream, block_path):
+        if not block_path:
+            if isinstance(stream, RichText):
+                yield from find_objects_in_rich_text(stream.source)
+            else:
+                yield stream
+            return
+        current_block, *block_path = block_path
+        if isinstance(current_block, StreamBlock) \
+                and isinstance(stream, StreamValue):
+            for sub_value in stream:
+                yield from self.find_values(sub_value, block_path)
+        elif isinstance(stream, StreamValue.StreamChild):
+            if stream.block == current_block:
+                yield from self.find_values(stream.value, block_path)
+        elif isinstance(stream, StructValue):
+            if current_block.name in stream:
+                yield from self.find_values(stream[current_block.name],
+                                            block_path)
+        elif current_block.name == '' and isinstance(stream, list):
+            for sub_value in stream:
+                yield from self.find_values(sub_value, block_path)
+        else:
+            warn('Unexpected StreamField value: <%s: %s>'
+                 % (type(stream), str(stream)[:30]))
+
+
+class ModelStreamFieldsCollector:
+    def __init__(self, model, using=DEFAULT_DB_ALIAS):
+        self.model = model
+        self.using = using
+        self.fields = [field for field in model._meta.fields
+                       if isinstance(field, StreamField)]
+        self.field_collectors = [StreamFieldCollector(field)
+                                 for field in self.fields]
+
+    def prepare_value(self, value):
+        if isinstance(value, Model):
```

Comment:
(Reposting this comment as I made it against a commit on my own branch rather than this PR, and not sure if you got the notification, @BertrandBordage...)

Do we ever call `find_objects` with values that aren't model instances? If so, it looks like we're missing tests for these cases, and we should also consider renaming `find_objects` to `find_values`. Personally I'd prefer to remove this functionality if there's no immediate need for it, though - it seems like feature creep, and adds complexity that we could do without. (I'm not entirely clear on why `"` and `|` characters in strings need to be double-escaped, or why we can't replace most of this method with `json.dumps`, and if we can avoid going down that rabbit-hole, we should :-) )

Reply:
Right, so would you be OK with you if that simplification consisted in just removing `prepare_value` and just systematically apply `json.dumps(v.pk).replace('|', r'\|')` in `find_objects`?

Sorry I don’t commit anymore on this branch, I’m not quite sure what to do now that you’ve started your own ^^

Reply:
Yep, I'll go ahead and remove `prepare_value` and any other code paths that become redundant as a result (I think there are a few `if isinstance(value, Model):` switches elsewhere).

I still don't understand why you're manually escaping `|` _in addition to_ calling `re.escape`. Again I'm struggling with the lack of docstring here - is the return value of `prepare_value` intended to be a literal string to match within the json (in which case it's the caller's responsibility to escape that literal string appropriately for use in a regexp, or whatever string-matching mechanism it decides to use), or a regexp pattern (in which case we should call `re.replace` on the literal string, since `|` is not the only character that has special meaning)?

Reply:
Oh, you’re right, the `|` escaping must be a mistake given that I added an `re.escape`. The only purpose of this method was to prepare the value before injecting it in `stream_structure_pattern`, so it had to be escaped so it’s considered as a literal by `stream_structure_pattern`.